File: objc/main/tests/output/TestRegions.m
#import "TestRegions.h"

@implementation TestRegions

- (void) testEmbeddedRegion {
  NSString * dir = [self randomDir];
  NSString * groupFile = [[@"a() ::= <<\n" stringByAppendingString:@"[<@r>bar<@end>]\n"] stringByAppendingString:@">>\n"];
  [self writeFile:dir param1:@"group.stg" param2:groupFile];
  STGroup * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/group.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  NSString * expected = @"[bar]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testRegion {
  NSString * dir = [self randomDir];
  NSString * groupFile = [[@"a() ::= <<\n" stringByAppendingString:@"[<@r()>]\n"] stringByAppendingString:@">>\n"];
  [self writeFile:dir param1:@"group.stg" param2:groupFile];
  STGroup * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/group.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  NSString * expected = @"[]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testDefineRegionInSubgroup {
  NSString * dir = [self randomDir];
  NSString * g1 = @"a() ::= <<[<@r()>]>>\n";
  [self writeFile:dir param1:@"g1.stg" param2:g1];
  NSString * g2 = @"@a.r() ::= <<foo>>\n";
  [self writeFile:dir param1:@"g2.stg" param2:g2];
  STGroup * group1 = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g1.stg"]] autorelease];
  STGroup * group2 = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g2.stg"]] autorelease];
  [group2 importTemplates:group1];
  ST * st = [group2 getInstanceOf:@"a"];
  NSString * expected = @"[foo]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testDefineRegionInSubgroupThatRefsSuper {
  NSString * dir = [self randomDir];
  NSString * g1 = @"a() ::= <<[<@r>foo<@end>]>>\n";
  [self writeFile:dir param1:@"g1.stg" param2:g1];
  NSString * g2 = @"@a.r() ::= <<(<@super.r()>)>>\n";
  [self writeFile:dir param1:@"g2.stg" param2:g2];
  STGroup * group1 = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g1.stg"]] autorelease];
  STGroup * group2 = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g2.stg"]] autorelease];
  [group2 importTemplates:group1];
  ST * st = [group2 getInstanceOf:@"a"];
  NSString * expected = @"[(foo)]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testDefineRegionInSubgroup2 {
  NSString * dir = [self randomDir];
  NSString * g1 = @"a() ::= <<[<@r()>]>>\n";
  [self writeFile:dir param1:@"g1.stg" param2:g1];
  NSString * g2 = @"@a.r() ::= <<foo>>>\n";
  [self writeFile:dir param1:@"g2.stg" param2:g2];
  STGroup * group1 = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g1.stg"]] autorelease];
  STGroup * group2 = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g2.stg"]] autorelease];
  [group1 importTemplates:group2];
  ST * st = [group1 getInstanceOf:@"a"];
  NSString * expected = @"[]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testDefineRegionInSameGroup {
  NSString * dir = [self randomDir];
  NSString * g = [@"a() ::= <<[<@r()>]>>\n" stringByAppendingString:@"@a.r() ::= <<foo>>\n"];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroup * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  NSString * expected = @"[foo]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testAnonymousTemplateInRegion {
  NSString * dir = [self randomDir];
  NSString * g = [[[@"a() ::= <<[<@r()>]>>\n" stringByAppendingString:@"@a.r() ::= <<\n"] stringByAppendingString:@"<[\"foo\"]:{x|<x>}>\n"] stringByAppendingString:@">>\n"];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroup * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  NSString * expected = @"[foo]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testCantDefineEmbeddedRegionAgain {
  NSString * dir = [self randomDir];
  NSString * g = [@"a() ::= <<[<@r>foo<@end>]>>\n" stringByAppendingString:@"@a.r() ::= <<bar>>\n"];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  ErrorBuffer * errors = [[[ErrorBuffer alloc] init] autorelease];
  [group setListener:errors];
  [group load];
  NSString * expected = [@"g.stg 2:3: region a.r is embedded and thus already implicitly defined" stringByAppendingString:newline];
  NSString * result = [errors description];
  [self assertEquals:expected param1:result];
}

- (void) testCantDefineEmbeddedRegionAgainInTemplate {
  NSString * dir = [self randomDir];
  NSString * g = [[[[[@"a() ::= <<\n" stringByAppendingString:@"[\n"] stringByAppendingString:@"<@r>foo<@end>\n"] stringByAppendingString:@"<@r()>"] stringByAppendingString:@"]\n"] stringByAppendingString:@">>\n"];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  ErrorBuffer * errors = [[[ErrorBuffer alloc] init] autorelease];
  [group setListener:errors];
  [group load];
  NSString * expected = [@"g.stg 3:2: redefinition of region a.r" stringByAppendingString:newline];
  NSString * result = [errors description];
  [self assertEquals:expected param1:result];
}

- (void) testMissingRegionName {
  NSString * dir = [self randomDir];
  NSString * g = @"@t.() ::= \"\"\n";
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  ErrorBuffer * errors = [[[ErrorBuffer alloc] init] autorelease];
  [group setListener:errors];
  [group load];
  NSString * expected = [@"g.stg 1:3: missing ID at '('" stringByAppendingString:newline];
  NSString * result = [errors description];
  [self assertEquals:expected param1:result];
}

- (void) testIndentBeforeRegionIsIgnored {
  NSString * dir = [self randomDir];
  NSString * g = [[[[@"a() ::= <<[\n" stringByAppendingString:@"  <@r>\n"] stringByAppendingString:@"  foo\n"] stringByAppendingString:@"  <@end>\n"] stringByAppendingString:@"]>>\n"];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  NSString * expected = [[@"[\n" stringByAppendingString:@"  foo\n"] stringByAppendingString:@"]"];
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testRegionOverrideStripsNewlines {
  NSString * dir = [self randomDir];
  NSString * g = [[[@"a() ::= \"X<@r()>Y\"" stringByAppendingString:@"@a.r() ::= <<\n"] stringByAppendingString:@"foo\n"] stringByAppendingString:@">>\n"];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  NSString * sub = [@"@a.r() ::= \"A<@super.r()>B\"" stringByAppendingString:newline];
  [self writeFile:dir param1:@"sub.stg" param2:sub];
  STGroupFile * subGroup = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/sub.stg"]] autorelease];
  [subGroup importTemplates:group];
  ST * st = [subGroup getInstanceOf:@"a"];
  NSString * result = [st render];
  NSString * expecting = @"XAfooBY";
  [self assertEquals:expecting param1:result];
}

- (void) testRegionOverrideRefSuperRegion {
  NSString * dir = [self randomDir];
  NSString * g = [@"a() ::= \"X<@r()>Y\"" stringByAppendingString:@"@a.r() ::= \"foo\""] + newline;
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  NSString * sub = [@"@a.r() ::= \"A<@super.r()>B\"" stringByAppendingString:newline];
  [self writeFile:dir param1:@"sub.stg" param2:sub];
  STGroupFile * subGroup = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/sub.stg"]] autorelease];
  [subGroup importTemplates:group];
  ST * st = [subGroup getInstanceOf:@"a"];
  NSString * result = [st render];
  NSString * expecting = @"XAfooBY";
  [self assertEquals:expecting param1:result];
}

- (void) testRegionOverrideRefSuperRegion3Levels {
  NSString * dir = [self randomDir];
  NSString * g = [@"a() ::= \"X<@r()>Y\"" stringByAppendingString:@"@a.r() ::= \"foo\""] + newline;
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  NSString * sub = [@"@a.r() ::= \"<@super.r()>2\"" stringByAppendingString:newline];
  [self writeFile:dir param1:@"sub.stg" param2:sub];
  STGroupFile * subGroup = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/sub.stg"]] autorelease];
  [subGroup importTemplates:group];
  NSString * subsub = [@"@a.r() ::= \"<@super.r()>3\"" stringByAppendingString:newline];
  [self writeFile:dir param1:@"subsub.stg" param2:subsub];
  STGroupFile * subSubGroup = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/subsub.stg"]] autorelease];
  [subSubGroup importTemplates:subGroup];
  ST * st = [subSubGroup getInstanceOf:@"a"];
  NSString * result = [st render];
  NSString * expecting = @"Xfoo23Y";
  [self assertEquals:expecting param1:result];
}

- (void) testRegionOverrideRefSuperImplicitRegion {
  NSString * dir = [self randomDir];
  NSString * g = [@"a() ::= \"X<@r>foo<@end>Y\"" stringByAppendingString:newline];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  NSString * sub = [@"@a.r() ::= \"A<@super.r()>\"" stringByAppendingString:newline];
  [self writeFile:dir param1:@"sub.stg" param2:sub];
  STGroupFile * subGroup = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/sub.stg"]] autorelease];
  [subGroup importTemplates:group];
  ST * st = [subGroup getInstanceOf:@"a"];
  NSString * result = [st render];
  NSString * expecting = @"XAfooY";
  [self assertEquals:expecting param1:result];
}

- (void) testUnknownRegionDefError {
  NSString * dir = [self randomDir];
  NSString * g = [[[@"a() ::= <<\n" stringByAppendingString:@"X<@r()>Y"] stringByAppendingString:@">>\n"] stringByAppendingString:@"@a.q() ::= \"foo\""] + newline;
  STErrorListener * errors = [[[ErrorBuffer alloc] init] autorelease];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  [group setListener:errors];
  ST * st = [group getInstanceOf:@"a"];
  [st render];
  NSString * result = [errors description];
  NSString * expecting = [@"g.stg 4:3: template a doesn't have a region called q" stringByAppendingString:newline];
  [self assertEquals:expecting param1:result];
}

- (void) testSuperRegionRefMissingOk {
  NSString * dir = [self randomDir];
  NSString * g = [@"a() ::= \"X<@r()>Y\"" stringByAppendingString:@"@a.r() ::= \"foo\""] + newline;
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroupFile * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/g.stg"]] autorelease];
  NSString * sub = [@"@a.r() ::= \"A<@super.q()>B\"" stringByAppendingString:newline];
  STErrorListener * errors = [[[ErrorBuffer alloc] init] autorelease];
  [group setListener:errors];
  [self writeFile:dir param1:@"sub.stg" param2:sub];
  STGroupFile * subGroup = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/sub.stg"]] autorelease];
  [subGroup importTemplates:group];
  ST * st = [subGroup getInstanceOf:@"a"];
  NSString * result = [st render];
  NSString * expecting = @"XABY";
  [self assertEquals:expecting param1:result];
}

- (void) testEmbeddedRegionOnOneLine {
  NSString * dir = [self randomDir];
  NSString * groupFile = [[[[@"a() ::= <<\n" stringByAppendingString:@"[\n"] stringByAppendingString:@"  <@r>bar<@end>\n"] stringByAppendingString:@"]\n"] stringByAppendingString:@">>\n"];
  [self writeFile:dir param1:@"group.stg" param2:groupFile];
  STGroup * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/group.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  [st.impl dump];
  NSString * expected = [[[@"[" stringByAppendingString:newline] stringByAppendingString:@"  bar"] + newline stringByAppendingString:@"]"];
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testEmbeddedRegionTagsOnSeparateLines {
  NSString * dir = [self randomDir];
  NSString * groupFile = [[[[[[@"a() ::= <<\n" stringByAppendingString:@"[\n"] stringByAppendingString:@"  <@r>\n"] stringByAppendingString:@"  bar\n"] stringByAppendingString:@"  <@end>\n"] stringByAppendingString:@"]\n"] stringByAppendingString:@">>\n"];
  [self writeFile:dir param1:@"group.stg" param2:groupFile];
  STGroup * group = [[[STGroupFile alloc] init:[dir stringByAppendingString:@"/group.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  NSString * expected = [[[@"[" stringByAppendingString:newline] stringByAppendingString:@"  bar"] + newline stringByAppendingString:@"]"];
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

@end

File: test/org/stringtemplate/v4/test/TestRegions.java
/*
 [The "BSD license"]
 Copyright (c) 2009 Terence Parr
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package org.stringtemplate.v4.test;

import org.junit.*;
import org.stringtemplate.v4.*;
import org.stringtemplate.v4.misc.ErrorBuffer;

import static org.junit.Assert.assertEquals;

public class TestRegions extends BaseTest {
    @Test public void testEmbeddedRegion()
    {
        String dir = getRandomDir();
        String groupFile =
            "a() ::= <<\n" +
            "[<@r>bar<@end>]\n" +
            ">>\n";
        writeFile(dir, "group.stg", groupFile);
        STGroup group = new STGroupFile(dir+"/group.stg");
        ST st = group.getInstanceOf("a");
        String expected = "[bar]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testRegion()
    {
        String dir = getRandomDir();
        String groupFile =
            "a() ::= <<\n" +
            "[<@r()>]\n" +
            ">>\n";
        writeFile(dir, "group.stg", groupFile);
        STGroup group = new STGroupFile(dir+"/group.stg");
        ST st = group.getInstanceOf("a");
        String expected = "[]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testDefineRegionInSubgroup()
    {
        String dir = getRandomDir();
        writeFile(dir, "g1.stg", "a() ::= <<[<@r()>]>>\n");
        writeFile(dir, "g2.stg", "@a.r() ::= <<foo>>\n");

        STGroup group1 = new STGroupFile(dir+"/g1.stg");
        STGroup group2 = new STGroupFile(dir+"/g2.stg");
        group2.importTemplates(group1); // define r in g2
        ST st = group2.getInstanceOf("a");
        String expected = "[foo]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testDefineRegionInSubgroupOneInSubdir()
    {
        String dir = getRandomDir();
        writeFile(dir, "g1.stg", "a() ::= <<[<@r()>]>>\n");
        writeFile(dir+"/subdir", "g2.stg", "@a.r() ::= <<foo>>\n");

        STGroup group1 = new STGroupFile(dir+"/g1.stg");
        STGroup group2 = new STGroupFile(dir+"/subdir/g2.stg");
        group2.importTemplates(group1); // define r in g2
        ST st = group2.getInstanceOf("a");
        String expected = "[foo]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testDefineRegionInSubgroupBothInSubdir()
    {
        String dir = getRandomDir();
        writeFile(dir+"/subdir", "g1.stg", "a() ::= <<[<@r()>]>>\n");
        writeFile(dir+"/subdir", "g2.stg", "@a.r() ::= <<foo>>\n");

        STGroup group1 = new STGroupFile(dir+"/subdir/g1.stg");
        STGroup group2 = new STGroupFile(dir+"/subdir/g2.stg");
        group2.importTemplates(group1); // define r in g2
        ST st = group2.getInstanceOf("a");
        String expected = "[foo]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testDefineRegionInSubgroupThatRefsSuper()
    {
        String dir = getRandomDir();
        String g1 = "a() ::= <<[<@r>foo<@end>]>>\n";
        writeFile(dir, "g1.stg", g1);
        String g2 = "@a.r() ::= <<(<@super.r()>)>>\n";
        writeFile(dir, "g2.stg", g2);

        STGroup group1 = new STGroupFile(dir+"/g1.stg");
        STGroup group2 = new STGroupFile(dir+"/g2.stg");
        group2.importTemplates(group1); // define r in g2
        ST st = group2.getInstanceOf("a");
        String expected = "[(foo)]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testDefineRegionInSubgroup2()
    {
        String dir = getRandomDir();
        String g1 = "a() ::= <<[<@r()>]>>\n";
        writeFile(dir, "g1.stg", g1);
        String g2 = "@a.r() ::= <<foo>>>\n";
        writeFile(dir, "g2.stg", g2);

        STGroup group1 = new STGroupFile(dir+"/g1.stg");
        STGroup group2 = new STGroupFile(dir+"/g2.stg");
        group1.importTemplates(group2); // opposite of previous; g1 imports g2
        ST st = group1.getInstanceOf("a");
        String expected = "[]"; // @a.r implicitly defined in g1; can't see g2's
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testDefineRegionInSameGroup()
    {
        String dir = getRandomDir();
        String g = "a() ::= <<[<@r()>]>>\n"+
                   "@a.r() ::= <<foo>>\n";
        writeFile(dir, "g.stg", g);

        STGroup group = new STGroupFile(dir+"/g.stg");
        ST st = group.getInstanceOf("a");
        String expected = "[foo]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testAnonymousTemplateInRegion()
    {
        String dir = getRandomDir();
        String g = "a() ::= <<[<@r()>]>>\n" +
                   "@a.r() ::= <<\n" +
                   "<[\"foo\"]:{x|<x>}>\n" +
                   ">>\n";
        writeFile(dir, "g.stg", g);

        STGroup group = new STGroupFile(dir+"/g.stg");
        ST st = group.getInstanceOf("a");
        String expected = "[foo]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testAnonymousTemplateInRegionInSubdir()
    {
        //fails since it makes region name /region__/g/a/_r
        String dir = getRandomDir();
        String g = "a() ::= <<[<@r()>]>>\n" +
                   "@a.r() ::= <<\n" +
                   "<[\"foo\"]:{x|<x>}>\n" +
                   ">>\n";
        writeFile(dir, "g.stg", g);

        STGroup.verbose = true;
        STGroup group = new STGroupDir(dir);
        ST st = group.getInstanceOf("g/a");
        String expected = "[foo]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testCantDefineEmbeddedRegionAgain()
    {
        String dir = getRandomDir();
        String g = "a() ::= <<[<@r>foo<@end>]>>\n"+
                   "@a.r() ::= <<bar>>\n"; // error; dup
        writeFile(dir, "g.stg", g);

        STGroupFile group = new STGroupFile(dir+"/g.stg");
        ErrorBuffer errors = new ErrorBuffer();
        group.setListener(errors);
        group.load();
        String expected = "g.stg 2:3: region /a.r is embedded and thus already implicitly defined"+newline;
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testCantDefineEmbeddedRegionAgainInTemplate()
    {
        String dir = getRandomDir();
        String g =
            "a() ::= <<\n" +
            "[\n" +
            "<@r>foo<@end>\n" +
            "<@r()>" +
            "]\n" +
            ">>\n"; // error; dup
        writeFile(dir, "g.stg", g);

        STGroupFile group = new STGroupFile(dir+"/g.stg");
        ErrorBuffer errors = new ErrorBuffer();
        group.setListener(errors);
        group.load();
        String expected = "g.stg 3:2: redefinition of region /a.r"+newline;
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testMissingRegionName()
    {
        String dir = getRandomDir();
        String g = "@t.() ::= \"\"\n";
        writeFile(dir, "g.stg", g);

        STGroupFile group = new STGroupFile(dir+"/g.stg");
        ErrorBuffer errors = new ErrorBuffer();
        group.setListener(errors);
        group.load();
        String expected = "g.stg 1:3: missing ID at '('"+newline;
        String result = errors.toString();
        assertEquals(expected, result);
    }

    @Test public void testIndentBeforeRegionIsIgnored()
    {
        String dir = getRandomDir();
        String g = "a() ::= <<[\n" +
                   "  <@r>\n" +
                   "  foo\n" +
                   "  <@end>\n" +
                   "]>>\n";
        writeFile(dir, "g.stg", g);

        STGroupFile group = new STGroupFile(dir+"/g.stg");
        ST st = group.getInstanceOf("a");
        String expected = "[" + newline +
                          "  foo" + newline +
                          "]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testRegionOverrideStripsNewlines()
    {
        String dir = getRandomDir();
        String g =
                "a() ::= \"X<@r()>Y\"" +
                "@a.r() ::= <<\n" +
                "foo\n" +
                ">>\n";
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
                "@a.r() ::= \"A<@super.r()>B\"" +newline;
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        ST st = subGroup.getInstanceOf("a");
        String result = st.render();
        String expecting = "XAfooBY";
        assertEquals(expecting, result);
    }

    //

    @Test public void testRegionOverrideRefSuperRegion()
    {
        String dir = getRandomDir();
        String g =
                "a() ::= \"X<@r()>Y\"" +
                "@a.r() ::= \"foo\"" +newline;
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
                "@a.r() ::= \"A<@super.r()>B\"" +newline;
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        ST st = subGroup.getInstanceOf("a");
        String result = st.render();
        String expecting = "XAfooBY";
        assertEquals(expecting, result);
    }

    @Test public void testRegionOverrideRefSuperRegion2Levels()
    {
        String g =
                "a() ::= \"X<@r()>Y\"\n" +
                "@a.r() ::= \"foo\"\n";
        STGroup group = new STGroupString(g);

        String sub =
                "@a.r() ::= \"<@super.r()>2\"\n";
        STGroup subGroup = new STGroupString(sub);
        subGroup.importTemplates(group);

        ST st = subGroup.getInstanceOf("a");

        String result = st.render();
        String expecting = "Xfoo2Y";
        assertEquals(expecting, result);
    }

    @Test public void testRegionOverrideRefSuperRegion3Levels()
    {
        String dir = getRandomDir();
        String g =
                "a() ::= \"X<@r()>Y\"" +
                "@a.r() ::= \"foo\"" +newline;
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
                "@a.r() ::= \"<@super.r()>2\"" +newline;
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        String subsub =
                "@a.r() ::= \"<@super.r()>3\"" +newline;
        writeFile(dir, "subsub.stg", subsub);
        STGroupFile subSubGroup = new STGroupFile(dir+"/subsub.stg");
        subSubGroup.importTemplates(subGroup);

        ST st = subSubGroup.getInstanceOf("a");

        String result = st.render();
        String expecting = "Xfoo23Y";
        assertEquals(expecting, result);
    }

    @Test public void testRegionOverrideRefSuperImplicitRegion()
    {
        String dir = getRandomDir();
        String g =
                "a() ::= \"X<@r>foo<@end>Y\""+newline;
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
                "@a.r() ::= \"A<@super.r()>\"" +newline;
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        ST st = subGroup.getInstanceOf("a");
        String result = st.render();
        String expecting = "XAfooY";
        assertEquals(expecting, result);
    }

    @Test public void testUnknownRegionDefError()
    {
        String dir = getRandomDir();
        String g =
                "a() ::= <<\n" +
                "X<@r()>Y\n" +
                ">>\n"+
                "@a.q() ::= \"foo\"" +newline;
        STErrorListener errors = new ErrorBuffer();
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("a");
        st.render();
        String result = errors.toString();
        String expecting = "g.stg 4:3: template /a doesn't have a region called q"+newline;
        assertEquals(expecting, result);
    }

    @Test public void testSuperRegionRefMissingOk()
    {
        String dir = getRandomDir();
        String g =
            "a() ::= \"X<@r()>Y\"" +
            "@a.r() ::= \"foo\"" +newline;
        writeFile(dir, "g.stg", g);
        STGroupFile group = new STGroupFile(dir+"/g.stg");

        String sub =
            "@a.r() ::= \"A<@super.q()>B\"" +newline; // allow this; trap at runtime
        STErrorListener errors = new ErrorBuffer();
        group.setListener(errors);
        writeFile(dir, "sub.stg", sub);
        STGroupFile subGroup = new STGroupFile(dir+"/sub.stg");
        subGroup.importTemplates(group);

        ST st = subGroup.getInstanceOf("a");
        String result = st.render();
        String expecting = "XABY";
        assertEquals(expecting, result);
    }

    @Test public void testEmbeddedRegionOnOneLine()
    {
        String dir = getRandomDir();
        String groupFile =
            "a() ::= <<\n" +
            "[\n" +
            "  <@r>bar<@end>\n" +
            "]\n" +
            ">>\n";
        writeFile(dir, "group.stg", groupFile);
        STGroup group = new STGroupFile(dir+"/group.stg");
        ST st = group.getInstanceOf("a");
        st.impl.dump();
        String expected = "["+newline+"  bar"+newline+"]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testEmbeddedRegionTagsOnSeparateLines()
    {
        String dir = getRandomDir();
        String groupFile =
            "a() ::= <<\n" +
            "[\n" +
            "  <@r>\n" +
            "  bar\n" +
            "  <@end>\n" +
            "]\n" +
            ">>\n";
        writeFile(dir, "group.stg", groupFile);
        STGroup group = new STGroupFile(dir+"/group.stg");
        ST st = group.getInstanceOf("a");
        String expected = "["+newline+"  bar"+newline+"]";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Ignore("will revisit the behavior of indented expressions spanning multiple lines for a future release")
    @Test public void testEmbeddedSubtemplate()
    {
        String dir = getRandomDir();
        String groupFile =
            "a() ::= <<\n" +
            "[\n" +
            "  <{\n" +
            "  bar\n" +
            "  }>\n" +
            "]\n" +
            ">>\n";
        writeFile(dir, "group.stg", groupFile);
        STGroup group = new STGroupFile(dir+"/group.stg");
        ST st = group.getInstanceOf("a");
        String expected = "["+newline+"  bar"+newline+"]";
        String result = st.render();
        assertEquals(expected, result);
    }
}
